sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisReader.java

```diff
@@ -142,4 +147,26 @@ public Instant getWatermark() {
     return source;
   }
 
+  /**
+   * Returns total size of all records that remain in Kinesis stream after current watermark.
+   * When currently processed record is not further behind than {@link #allowedRecordLateness}
+   * then this method returns 0.
+   */
+  @Override
+  public long getTotalBacklogBytes() {
+    Instant watermark = getWatermark();
+    if (watermark.plus(allowedRecordLateness).isAfterNow()) {
+      return 0L;
+    }
+    long backlogBytes = lastBacklogBytes;
+    try {
+      backlogBytes = kinesis.getBacklogBytes(source.getStreamName(), watermark);
```

Comment:
I would think so, but you can postpone it with a comment noting about the performance. With `backlogThreshold` (lateness), this might be mute when the pipeline is upto date, but the extra latency penalty kicks in when the pipeline is already has high backlog, which can exacerbate the problem that caused the delay in the first place.
Caching is requires just storing last backlog and request time in local transient state of the reader.